presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java

```diff
@@ -0,0 +1,282 @@
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitProject(ProjectNode node, RewriteContext<InsertSemiJoinContext> context)
+        {
+            InPredicate inPredicate = context.get().peekInPredicate();
+            InPredicate unprojectedInPredicate = unprojectInPredicate(node, inPredicate);
+            if (!inPredicate.equals(unprojectedInPredicate)) {
+                context.get().pushInPredicate(unprojectedInPredicate);
+            }
+            node = (ProjectNode) super.visitProject(node, context);
+            Optional<Symbol> inPredicateSymbol = context.get().getInPredicateSymbol();
+            if (inPredicateSymbol.isPresent()) {
+                return tryAppendProjection(node, inPredicateSymbol.get());
+            }
+            else {
+                return node;
+            }
+        }
+
+        private InPredicate unprojectInPredicate(ProjectNode node, InPredicate inPredicate)
+        {
+            Expression value = inPredicate.getValue();
+            Expression valueList = inPredicate.getValueList();
+
+            for (Symbol assignmentKey : node.getAssignments().keySet()) {
+                Expression assignmentValue = node.getAssignments().get(assignmentKey);
+                Expression assigmentKeyExpression = assignmentKey.toQualifiedNameReference();
+                if (value.equals(assigmentKeyExpression)) {
+                    value = assignmentValue;
+                }
+                if (valueList.equals(assigmentKeyExpression)) {
+                    valueList = assignmentValue;
+                }
+                inPredicate = new InPredicate(value, valueList);
+            }
+            return inPredicate;
+        }
+
+        private ProjectNode tryAppendProjection(ProjectNode node, Symbol symbol)
+        {
+            if (node.getOutputSymbols().contains(symbol)) {
+                return node;
+            }
+            else if (node.getSource().getOutputSymbols().contains(symbol)) {
+                ImmutableMap.Builder<Symbol, Expression> builder = ImmutableMap.builder();
+                builder.putAll(node.getAssignments());
+                builder.put(symbol, symbol.toQualifiedNameReference());
+                return new ProjectNode(node.getId(), node.getSource(), builder.build());
+            }
+            else {
+                return node;
+            }
+        }
+
+        @Override
+        public PlanNode visitApply(ApplyNode node, RewriteContext<InsertSemiJoinContext> context)
+        {
+            InPredicate inPredicate = context.get().peekInPredicate();
+            if (inPredicate.getValueList() instanceof QualifiedNameReference) {
+                Symbol value = asSymbol(inPredicate.getValue());
```

Comment:
What if value is literal like in:
`SELECT 1 IN SELECT ....`
will that work - or throw in `asSymbol()` due to `1` not being `QualifiedNameReference`?